File: Java/src/rcr/robots/dagucar/DaguCar.java
package rcr.robots.dagucar;

import gnu.io.NoSuchPortException;
import gnu.io.PortInUseException;
import gnu.io.UnsupportedCommOperationException;

import java.io.IOException;

import rcr.utils.Serial;
import rcr.utils.SerialTimeoutException;

/**
 * Clase para interactuar con el rauto Dagucar/iRacer
 * Mayor información en https://www.sparkfun.com/products/11162.
 * En Chile lo comercializaba Olimex
 *
 * @author Roberto Carrasco
 */
 public class DaguCar {
    /** Detiene el auto */
    private static final int CMD_STOP = 0;
    /**  Mueve el auto hacia adelante */
    private static final int CMD_FORWARD = 1;
    /**  Mueve el auto hacia atrás */
    private static final int CMD_BACKWARD = 2;
    /**  Mueve las ruedas del auto hacia la izquierda */
    private static final int CMD_LEFT = 3;
    /**  Mueve las ruedas del auto hacia la derecha */
    private static final int CMD_RIGHT = 4;
    /**  Mueve el auto hacia adelante e izquierda */
    private static final int CMD_FORWARD_LEFT = 5;
    /**  Mueve el auto hacia adelante y derecha */
    private static final int CMD_FORWARD_RIGHT = 6;
    /**  Mueve el auto hacia atrás e izquierda */
    private static final int CMD_BACKWARD_LEFT = 7;
    /**  Mueve el auto hacia atrás y derecha */
    private static final int CMD_BACKWARD_RIGHT = 8;

    /** el objeto de la conexión serial */
    private Serial serial;
    /** el último comando ejecutado para evitar enviarlo dos veces */
    private int lastCmd = -1;

    /**
     * Construye el objeto para manipular el auto DaguCar/iRacer a través de la puerta serial
     *
     * @param port la puerta serial a la cual conectarse ("/dev/rfcomm1", "COM1")
     * @param timeout tiempo de espera en ms para recibir los datos desde el S2
     * @throws NoSuchPortException
     * @throws PortInUseException
     * @throws UnsupportedCommOperationException
     * @throws IOException
    */
    public DaguCar( String port, int timeout ) throws NoSuchPortException, PortInUseException, UnsupportedCommOperationException, IOException {
        for( int i = 1; ; i++ ) {
            try {
                serial = new Serial( port, 9600, timeout );
                Pause( 1000 );
                serial.FlushRead( 2000 );
                return;
            } catch( Exception e ){
                if( i > 10 ) {
                    throw e;
                }
            }
        }
    }

    /***
     * Cierra la conexión con el auto
     */
    public synchronized void Close() {
        serial.Close();
    }

    /**
     * Suspende la ejecución del thread por un número dado de milisegundos
     *
     * @param ms el número de ms a suspender el thread
     */
    public void Pause( int ms ) {
        serial.Pause( ms );
    }

    /*
     * Comandos para el auto
     */

    /***
     * Detiene el auto
     *
     * @throws IOException
     */
    public synchronized void Stop() throws IOException {
        Move( CMD_STOP, 0x00 );
    }

    /***
     * Mueve el auto hacia adelante
     *
     * @throws IOException
     */
    public synchronized void MoveForward( int speed ) throws IOException {
        Move( CMD_FORWARD, speed & 0x0F );
    }

    /***
     * Mueve el auto hacia atrás
     *
     * @throws IOException
     */
    public synchronized void MoveBackward( int speed ) throws IOException {
        Move( CMD_BACKWARD, speed & 0x0F );
    }

    /***
     * Mueve las ruedas del auto hacia la izquierda
     *
     * @throws IOException
     */
    public synchronized void MoveLeft( int speed ) throws IOException {
        Move( CMD_LEFT, speed & 0x0F );
    }

    /***
     * Mueve las ruedas del auto hacia la derecha
     *
     * @throws IOException
     */
    public synchronized void MoveRight( int speed ) throws IOException {
        Move( CMD_RIGHT, speed & 0x0F );
    }

    /***
     * Mueve el auto hacia adelante e izquierda
     *
     * @throws IOException
     */
    public synchronized void MoveForwardLeft( int speed ) throws IOException {
        Move( CMD_FORWARD_LEFT, speed & 0x0F );
    }

    /***
     * Mueve el auto hacia adelante y derecha
     *
     * @throws IOException
     */
    public synchronized void MoveForwardRight( int speed ) throws IOException {
        Move( CMD_FORWARD_RIGHT, speed & 0x0F );
    }

    /***
     * Mueve el auto hacia atrás e izquierda
     *
     * @throws IOException
     */
    public synchronized void MoveBackwardLeft( int speed ) throws IOException {
        Move( CMD_BACKWARD_LEFT, speed & 0x0F );
    }

    /***
     * Mueve el auto hacia atrás y derecha
     *
     * @throws IOException
     */
    public synchronized void MoveBackwardRight( int speed ) throws IOException {
        Move( CMD_BACKWARD_RIGHT, speed & 0x0F );
    }

    /**
     * métodos privados
     */

    private void Debug( String msg ) {
        System.out.println( msg );
    }

    private void Debug( int n ) {
        System.out.println( n );
    }

    private void Move( int direction, int speed ) throws IOException {
        if( direction < CMD_STOP || direction > CMD_BACKWARD_RIGHT ) {
            direction =CMD_STOP;
        }
        speed = speed | 0x0F;

        int cmd = ( direction<<4 ) | speed;
        if( cmd != lastCmd ) {
            byte[] packet = new byte[1];
            packet[0] = (byte)cmd;
            serial.Write( packet );
            lastCmd = cmd;
        }
    }

}

File: Python/rcr/robots/dagucar/DaguCar.py
# -*- coding: utf-8 -*-

from rcr.utils.Serial import Serial

import threading
import time

class DaguCar:
    CMD_STOP           = 0
    CMD_FORWARD        = 1
    CMD_BACKWARD       = 2
    CMD_LEFT           = 3
    CMD_RIGHT          = 4
    CMD_FORWARD_LEFT   = 5
    CMD_FORWARD_RIGHT  = 6
    CMD_BACKWARD_LEFT  = 7
    CMD_BACKWARD_RIGHT = 8

    def __init__( self, port, timeout ):
        self._lastCmd = -1
        self._lock = threading.Lock()
        for i in range( 20 ):
            try:
                self._ser = Serial(port, 9600, timeout)
                self.Pause( 1000 )
                self._ser.FlushRead( 2.0 )
                return
            except Exception as e:
                if( i > 10 ):
                    raise

    def Close( self ):
        try:
            self._Lock()
            self._ser.Close()
        except Exception as e:
            raise
        finally:
            self._Unlock()

    def Pause( self, ms ):
        self._ser.Pause( ms )

    ###
    # Comandos para el auto
    ###


    def Stop( self ):
        try:
            self._Lock()
            self._Move( self.CMD_STOP, 0x00 )
        except Exception as e:
            raise
        finally:
            self._Unlock()

    def MoveForward( self, speed ):
        try:
            self._Lock()
            self._Move( self.CMD_FORWARD, speed & 0x0F )
        except Exception as e:
            raise
        finally:
            self._Unlock()

    def MoveBackward( self, speed ):
        try:
            self._Lock()
            self._Move( self.CMD_BACKWARD, speed & 0x0F )
        except Exception as e:
            raise
        finally:
            self._Unlock()

    def MoveLeft( self, speed ):
        try:
            self._Lock()
            self._Move( self.CMD_LEFT, speed & 0x0F )
        except Exception as e:
            raise
        finally:
            self._Unlock()

    def MoveRight( self, speed ):
        try:
            self._Lock()
            self._Move( self.CMD_RIGHT, speed & 0x0F )
        except Exception as e:
            raise
        finally:
            self._Unlock()

    def MoveForwardLeft( self, speed ):
        try:
            self._Lock()
            self._Move( self.CMD_FORWARD_LEFT, speed & 0x0F )
        except Exception as e:
            raise
        finally:
            self._Unlock()

    def MoveForwardRight( self, speed ):
        try:
            self._Lock()
            self._Move( self.CMD_FORWARD_RIGHT, speed & 0x0F )
        except Exception as e:
            raise
        finally:
            self._Unlock()

    def MoveBackwardLeft( self, speed ):
        try:
            self._Lock()
            self._Move( self.CMD_BACKWARD_LEFT, speed & 0x0F )
        except Exception as e:
            raise
        finally:
            self._Unlock()

    def MoveBackwardRight( self, speed ):
        try:
            self._Lock()
            self._Move( self.CMD_BACKWARD_RIGHT, speed & 0x0F )
        except Exception as e:
            raise
        finally:
            self._Unlock()

    ###
    # métodos privados
    ###

    def _Debug( self, val ):
        print(val)

    def _Move( self, direction, speed ):
        if( direction < self.CMD_STOP or direction > self.CMD_BACKWARD_RIGHT ):
            direction = self.CMD_STOP
        speed = speed & 0x0F

        cmd = ( direction<<4 ) | speed
        if( cmd != self._lastCmd ):
            packet = bytearray( 1 )
            packet[0] = cmd
            self._ser.Write( packet )
            self._lastCmd = cmd

    def _Lock( self ):
        """Obtiene acceso exclusivo."""
        self._lock.acquire()

    def _Unlock( self ):
        """Libera el acceso exclusivo."""
        self._lock.release()
